### ISSUE TITLE
Ambiguous whether link points to a collection or single resource

### ISSUE THREAD

**Issue by ethanresnick** [json-api/json-api#249]
In cases where referenced resources are fetched from an external url (i.e. we're not dealing with a compound document), and the uri template in the `links` keys is expanded using the primary resources' ids—rather than the id(s) of the referenced resources—it's ambiguous whether the document returned at that url will contain a single resource or a collection of resources.

For example, in the following response for `GET /mytype`, it's unclear whether each `mytype.prop` would ultimately hold a single resource or multiple:

```
{
  links: {
    "mytype.prop": {
      "href": "http://.../{mytype.id}",
      "type": "type2"
    }
  }
  mytype: [{
    "id": "3",
    "attrName": "val"
  }, 
  //...
  ]
} 
```

The same thing occurs if you use a resource-level `links` key instead of a top level one. Here's a response for `GET /mytype/3`

```
{
  mytype: {
    "id": "3",
    "attrName": "val",
    "links": {
      "prop": {
        "href": "http://.../3",
        "type": "type2"
      }
    }
  }
} 
```

My question is: is this a problem? (Seems like it might complicate client-side code a bit). And, if it is, what to do about it?

**Comment by hesalx** [comment]
Late answer, but I think the question is reasonable.

**tl;dr** What would your client do is you make such request alone? Where to know if it's a collection or a single resource?

From the REST perspective either collection or single resource is just "a resource". A good client should not relies on this difference. (A related resource may not be a JSON API resource at all. A client should determine it according to the media type of response.)

In some sense REST tells you to know what the resource actually is at the moment of the response, not at the moment of the request. Otherwise it means that one request depends of another (earlier one). REST stands for the opposite behaviour.

JSON API provides the unified format for collections and single resources. It's either the `"data"` object or the object named after the `"type"`. And the type of this object is either array or object. Which determines the resource as a collection or a single resource clearly.

Actually, there is the proposal for schemas #310, which, if accepted, would help to define additional properties of related resources in such special cases, despite it would deviates from REST a bit, if it reasonable urgent for you.

**Comment by ethanresnick** [comment]
Yeah, what I'm talking about here would be addressed well by schemas, and that's the real issue. Because, for some clients, it's useful to know in advance what kind of response you're going to get back (e.g. how Ember Data will mock a collection of models vs a single model before making the request).

**Comment by steveklabnik** [comment]
Sorry for taking so long to respond here :(

> My question is: is this a problem?

I don't think that it is. as @arantir mentions:

> From the REST perspective either collection or single resource is just "a resource".

**Comment by bintoro** [comment]
Whoa. The conflation of _REST resource_ and _JSON API resource_ rears its head again... The JSON API resource system is basically an object model. A related resource URL points either to a singular resource or a logical collection of resources. This is a static property dependent on the cardinality of the relationship. It's not "just a resource".

(**EDIT:** To make this point absolutely clear, in REST a resource is just a resource because REST, unlike JSON API, has no concepts such as "ID", "type", "attribute", "to-one", or "to-many". REST has nothing to do with the JSON API resource model.)

Currently, whether a link points to a collection can be inferred when the `id` member is present (string vs. array), but not when we only have a URL. This is not ideal, particularly since RC2 implies that IDs are intended for linkage within compound documents only.

> A related resource may not be a JSON API resource at all. A client should determine it according to the media type of response.

Not true according to the spec.
